server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -227,6 +245,147 @@ public void close() throws IOException {
         IOUtils.close(persistedState.get());
     }
 
+    // visible for testing
+    public boolean allPendingAsyncStatesWritten() {
+        final PersistedState ps = persistedState.get();
+        if (ps instanceof AsyncLucenePersistedState) {
+            return ((AsyncLucenePersistedState) ps).allPendingAsyncStatesWritten();
+        } else {
+            return true;
+        }
+    }
+
+    static class AsyncLucenePersistedState extends InMemoryPersistedState {
+
+        private static final Logger logger = LogManager.getLogger(AsyncLucenePersistedState.class);
+
+        static final String THREAD_NAME = "AsyncLucenePersistedState#updateTask";
+
+        private final EsThreadPoolExecutor threadPoolExecutor;
+        private final LucenePersistedState lucenePersistedState;
+
+        boolean newCurrentTermQueued = false;
+        boolean newStateQueued = false;
+
+        private final Object mutex = new Object();
+
+        AsyncLucenePersistedState(Settings settings, ThreadPool threadPool, PersistedClusterStateService.Writer persistenceWriter,
+                                  long currentTerm, ClusterState lastAcceptedState) throws IOException {
+            super(currentTerm, lastAcceptedState);
+            final String nodeName = Objects.requireNonNull(Node.NODE_NAME_SETTING.get(settings));
+            threadPoolExecutor = EsExecutors.newScaling(
```

Comment:
Could this executor have a bounded queue? AFAICT there'll only be ≤1 task.

Reply:
yes, fixed in f6ada35fe92